crates/ide-completion/src/completions/postfix.rs

```diff
@@ -123,6 +123,12 @@ pub(crate) fn complete_postfix(
     postfix_snippet("ref", "&expr", &format!("&{receiver_text}")).add_to(acc);
     postfix_snippet("refm", "&mut expr", &format!("&mut {receiver_text}")).add_to(acc);
 
+    let unsafe_completion_string = match dot_receiver {
+        ast::Expr::BlockExpr(_) => format!("unsafe {receiver_text}"),
+        _ => format!("unsafe {{ {receiver_text} }}"),
+    };
```

Comment:
This will happily change the block of things like `if true {}` into `if true unsafe {}` which is invalid syntax. So we should check that this `BlockExpr` is not a child of `IfExpr`, `WhileExpr`, `LoopExpr`, `ForExpr`, or the right hand side of a `let ... = .. else {}`

Reply:
You can easily check this by taking the parent of dot_receiver via `syntax_node().parent()` and then checking the `SyntaxKind` of that via `kind()` for the corresponding expressions I have listed.

The let statement case is a bit more tricky, you'll have to check whether the else expression exists there and whether that is the same as the block expression here.

Reply:
Yes, basically for those blocks we want the previous behavior.

Reply:
Actually, I added those case to the test and it passed without changing the code. I'm not sure why.
Btw, the `let` case was `let ... = unsafe { if .. else {} }`, not `unsafe { let ... }`.

Reply:
Oh apologies, the `let` one is regarding `let .. = .. else { .. }` not `let .. = if .. { ..} else { .. }` (the first one is its own kind of statement, usuall referred to as `let else`).

If those cases already pass then that's good, it might be handled at some other place already (adding tests for them is still a good thing)

Reply:
I added code to check whether `BlockExpr` is a child of `IfExpr`, `LoopExpr`, `WhileExpr` and `ForExpr`.


For the case `LetElse`. I used the following test
```
 let x = true else {panic!()}.unsafe
```
However, it did not trigger postfix completion.

After debugging, I found that its has `NameRefKind::Path`, therefore it did not run into the following code  to trigger postfix completion.
https://github.com/rust-lang/rust-analyzer/blob/7f1234492e3164f9688027278df7e915bc1d919c/crates/ide-completion/src/completions.rs#L671-L675

Am I doing thing wrong, or it is expected behavior?

Reply:
Ah that is probably parser recovery kicking in and throwing completion infra over. That's fine then, no need to special case it if it doesn't trigger like this (do keep the test though).